### ISSUE TITLE
Add suppressTransactions optional flag to DbContext.Database.Migrate methods

### ISSUE THREAD

**Issue by ErroneousFatality** [dotnet/efcore#30881]
A while ago [the default migration script generation behavior was changed](https://github.com/dotnet/efcore/issues/7681#issuecomment-665848100) to always wrap the migration code inside a transaction.
It can be changed when running the migrations through the command console or package manager console and passing the `--no-transactions` option.
It can not be changed when calling the [DbContext.Database.MigrateAsync](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.relationaldatabasefacadeextensions.migrateasync?view=efcore-7.0). There's no parameter to tell the Migrate method to not use transactions.

I'm using [YugabyteDB](https://docs.yugabyte.com/preview/drivers-orms/csharp/entityframework) which doesn't support DDL code inside transactions.
I like to have a DatabaseMigrator application which is run during CI/CD to create, migrate and seed all our databases.

Could we add an optional `bool` parameter (e.g. `suppressTransactions`) to the [DbContext.Database.MigrateAsync](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.relationaldatabasefacadeextensions.migrateasync?view=efcore-7.0) and [DbContext.Database.Migrate](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.relationaldatabasefacadeextensions.migrate?view=efcore-7.0) methods, which when set would mean that the generated migration code will not contain transactions.

**Comment by FranckPachot** [comment]
Yes, it would be great to have the choice of using auto-commit (by not issuing begin/commit) like with other databases that don't support transactional DDL. YugabyteDB is PostgreSQL-compatible but transactional DDL is complex in a scale-out architecture and is not supported yet. This means that DDL is executed as another transaction and do not see the DML result that was run before in a begin/commit block (this is tracked by https://github.com/yugabyte/yugabyte-db/issues/3109)

**Comment by roji** [comment]
This sounds like something that should be taken care of by the provider, rather than a user-facing API addition to Migrate(). In other words, the Npgsql EF provider could have a Yugabyte mode, where migrations don't run under transactions. Note that this may require some support from EF Core.

**Comment by FranckPachot** [comment]
I agree and transactional DDL is on the roadmap but that's easier in a monolithic database than in a scale-out one. 
I think that being able to disable it (i.e run autocommit statements) is useful even for PostgreSQL. When the migration involves large tables that have to be reorganized, it can be better to avoid long transactions.

**Comment by roji** [comment]
> I agree and transactional DDL is on the roadmap but that's easier in a monolithic database than in a scale-out one.

Yeah, I can imagine the complexity involved there.

> I think that being able to disable it (i.e run autocommit statements) is useful even for PostgreSQL. When the migration involves large tables that have to be reorganized, it can be better to avoid long transactions.

I'm not too sure about that - having a one-time long transaction for the migrations seems better than the risk of something failing in the middle and leaving everything completely inconsistent... But we'll discuss this and see what the team thinks.

**Comment by ajcvickers** [comment]
@ErroneousFatality We discussed this to determine, beyond the question of what a provider should do, whether or not it makes sense to add this functionality to the API. Part of this was understanding why `--no-transactions` exists, given that it is not a good idea to execute migrations without transactions. The answer is that we added it specifically for use when creating the script at the same time that we added the functionality to create transactions to the script itself. Previous to this, the guidance was to create a transaction before running the script and to commit the transaction at the end. But this meant that code that existed to do this when running a script would be broken (or, at best, use nested transactions) when running with the new script. So we added the flag to allow opting out of the breaking change, and _not_ because _using no transactions at all_ is a good idea.

Continuing this line of thought, you state that "It can be changed when running the migrations through the command console or package manager console and passing the --no-transactions option." This isn't actually how it works `no-transactions` can only be used with the `dotnet ef migrations script` command. There is no way to prevent `dotnet ef database update` from using a transaction when the provider has indicated that the command should use a transaction.

Given all this, we don't believe it is a good idea to add this to `Migrate` API.

**Comment by ErroneousFatality** [comment]
@ajcvickers I understand. Thank you for taking the time to discuss this in such great detail.

**Comment by roji** [comment]
Reopening because I do think it may make sense to allow the provider (as opposed to the user) to disable transactions for migrations. The Npgsql provider could have a compatibility mode triggering this behavior for Yugabyte.
